src/map.rs

```diff
@@ -0,0 +1,114 @@
+use std::{marker::PhantomData, sync::Arc};
+
+use crate::Service;
+
+pub trait MapService<S1: Service, S2: Service>: std::fmt::Debug + Send + Sync + 'static {
+    fn req_up(&self, req: S2::Req) -> S1::Req;
```

Comment:
The two type parameters are called S1 and S2, but the methods are called up and down. Not sure why from s2 to s1 is up. Maybe call the type parameters something else? Or maybe just have a comment about S1 being the outer service or something? Difficult to have meaningful names with typical type parameter names...

Reply:
I renamed the generics and the methods and added docs, should be much clearer now.